README.markdown

```diff
@@ -4,6 +4,26 @@ If you simply want to run a local copy of OneZoom, but not modify the code yours
 
 # OneZoom setup
 
+## Installing in a Docker Dev Container
+
+If you are using Visual Studio Code or another editor that supports [Dev Containers](https://containers.dev/), the easiest way to set up a full development environment is to use the included Dev Container configuration. This will automatically create two containers: one for development (_dev_), and another (_web_) with the MySQL database and production web server (nginx + uwsgi), derived from the Docker image mentioned above. Your source code will be mounted simultaneously into both containers: dev will let you run build commands and web will serve it as a web server you can access via port forwarding. You can also run your own server from the dev container by [running web2py.py directly](#starting-and-shutting-down-web2py) -- this is particularly useful for debugging.
+
+If you are using Visual Studio Code, perform the following steps (you will need to modify these for another editor):
+
+1. Follow the instructions at either https://hub.docker.com/r/onezoom/oztree or https://hub.docker.com/r/onezoom/oztree-complete to save a docker image with IUCN data.
+1. Follow the [VSCode instructions for installing Dev Container support](https://code.visualstudio.com/docs/devcontainers/containers#_installation).
+1. `git clone https://github.com/OneZoom/OZtree` into the directory of your choice. If using Windows, it is highly recommended to [clone on the WSL2 filesystem](https://code.visualstudio.com/remote/advancedcontainers/improve-performance#_store-your-source-code-in-the-wsl-2-filesystem-on-windows) both for performance reasons and to avoid permissions issues. If you wish to fork the repository and clone your fork, you will need to copy the tags from upstream, otherwise you will see build issues later. You can do this with `git fetch --tags upstream` followed by `git push --tags`.
+1. Open the cloned directory in VSCode.
+1. Create a `.env` file at the root of the project and add `WEB_IMAGE_NAME=onezoom/oztree-with-iucn`, changing the value to whatever image name you choose in step 1.
+1. Open the command palette and choose "Dev Containers: Reopen in Container".
+1. Open an integrated terminal and run the following: `cp /opt/web2py/applications/OZtree_original/private/appconfig.ini /opt/web2py/applications/OZtree/private/ && cp /opt/web2py/applications/OZtree_original/databases/*.table /opt/web2py/applications/OZtree/databases/`
+1. Run `npm install && grunt dev`. You will need to rerun `grunt dev` any time you make code changes.
+1. Visit http://localhost and the website should load! You can also run your own server from the dev container by [running web2py.py directly](#starting-and-shutting-down-web2py).
```

Comment:
I got to this point, but http://localhost/ doesn't work. Is this really supposed to be on port 80?

Reply:
What OS are you on? Can you check the Ports view in VSCode? Maybe it had to choose a different local port. Does the web2py.py server work for you?

Oddly I didn't see a devcontainer config option to forward to a different local port or I definitely would've done that. I guess I could try to change this in the web Dockerfile but it probably would require more copied code to maintain.

Reply:
I'm on Win 11 / wsl2.

Here is the port view, which only has 8000, which is the dev container:
![image](https://github.com/OneZoom/OZtree/assets/556238/2345d67d-d38b-4d4b-bed2-642140856078)

The web2py.py server runs fine on http://127.0.0.1:8000/ (after manually running it).

So the one in the web container should be running all the time without being started, right?

One thing I'm confused about: when I run a terminal in VSC, does that always go into the dev container? Is there a way to open one in the web container?

It's all pretty new to me, so I may very well be doing something dumb.

Reply:
What happens if you add port 80 to the forwardPort list in devcontainer.json? This could be due to timing differences between our machines that cause the port opening to slip past the autoforwarding system, and if so we should remove the nondeterminism. You will need to run Rebuild Container from the palette for these changes to take effect. 

There is unfortunately no way to have the terminal open in the web container from VSC. This is a known limitation that they wish to address. You can get a terminal manually using the docker interface. Fortunately this should generally not be needed except perhaps for debugging container issues like this one.

Reply:
And to confirm: you should not need to manually start the server on port 80.

Reply:
Yep, that did it! It would indeed be nice to forward to a different local port instead of 80, but that's good enough for now.

Side note: you can open a terminal to the web container in VSC by choosing 'Attach to running container' and choosing oztree-web-1, which opens a separate VSC instance.

Reply:
Great. And yeah, I was thinking of GitHub codespaces where you can't attach to the web container.

Reply:
I was able to remap the port in the docker-compose, e.g.

```yaml
  web:
    ports:
      - "8001:80"
```

It then runs locally on http://localhost:8001/

Reply:
Nice! I'll make that change. Completely undocumented :-( https://containers.dev/implementors/json_reference/

Reply:
I found that on https://github.com/orgs/community/discussions/10394#discussioncomment-7185886. But note that I'm referring to the docker-compose file, not devcontainer.json. So it makes sense that it's not documented there, but it is in https://docs.docker.com/compose/compose-file/compose-file-v3/.

Reply:
Ah of course, makes sense.

Reply:
Updated. I used port 8080 to align with https://hub.docker.com/r/onezoom/oztree recommendation. Interestingly, my machine still auto-forwards port 80 (in addition to 8080). I tried using the devcontainer setting to disable this, but it did nothing.

Reply:
Yes, I do see that as well. Puzzling, but I guess let's not block on that for now. It's mostly a problem if someone is already running a web server like IIS on port 80.